### PR TITLE
Improve docker image to decrease size

### PR DESCRIPTION
**Pull request by ssz1997**
Use alpine instead of ubuntu1604-customize to reduce docker image size from 2.8G to 2.41G. Add and modify commands to make everything works in alpine.

### CODE REVIEW

**Comment by apc999** [comment]
@ssz1997 @LuQQiu   with alpine, are we still able to run fuse related? e.g., I remember the original reason to move from Alpine to Customized ubuntu is to have libfuse working

**Comment by ssz1997** [comment]
> @ssz1997 @LuQQiu with alpine, are we still able to run fuse related? e.g., I remember the original reason to move from Alpine to Customized ubuntu is to have libfuse working
> 
> also cheyang/fuse2:ubuntu1604-customize has a some optimization in libfuse.

I manually tested the fuse mount/unmount and it worked. I also migrated the optimization here.

**Comment by LuQQiu** [comment]
@ssz1997 have you tested any file read write path? like write a file into fuse, see if it shows in alluxio namespace/UFS. Read some file from fuse, etc.

**Comment by ssz1997** [comment]
> @ssz1997 have you tested any file read write path? like write a file into fuse, see if it shows in alluxio namespace/UFS. Read some file from fuse, etc.

I tried creating directory inside the alluxio mounting path - it appears both in alluxio namespace and mount point. Also tried copy a local file to the mounting path, and it can be read both from alluxio namespace and the mount point.

**Comment by apc999** [comment]
@madanadit  do you have the context when we switch from alpine to ubuntu? I remember it is for libfuse, but it seems @ssz1997  is able to run fuse without using ubuntu

**Comment by madanadit** [comment]
That's what I remember as well; that libfuse wasn't available on alpine. Could have changed since.

**Comment by apc999** [comment]
thanks @ssz1997  for verifying it works! Good to know.
@LuQQiu , can you work with @ssz1997 to understand if the `MAX_IDLE_THREAD` macro still works ? That's the customization @cheyang added to the `cheyang/fuse2:ubuntu1604-customize`

**Comment by apc999** [comment]
> That's what I remember as well; that libfuse wasn't available on alpine. Could have changed since.

double checked the PR, we actually checkedout and installed libfuse on top of Alpine -- that's why it works.

File: integration/docker/Dockerfile [hunk 1]
```diff
@@ -32,10 +32,37 @@ RUN if [ ${ENABLE_DYNAMIC_USER} = "true" ] ; then \
...
+   git checkout fuse_2_9_5_customize_multi_threads_no_logging
+
+#ADD libfuse /libfuse
+
+WORKDIR "/libfuse"
+
+RUN bash makeconf.sh && \
+  ./configure && \
+  make -j8 && \
+  make install
```
Comment (apc999) [inline integration/docker/Dockerfile:58]: after libfuse is created, can we remove all libfuse related src code and compilation tools to further reduce the image size?

**Comment by maobaolong** [comment]
@ssz1997 Thanks for the work you did to decrease the size of docker image. I definitely agree with build a small size image for stable production env. But, before stable, we need a env with jdk tools like jps, jstack, we also need ping, vi, net-tools to helps us to debug unexpected behavior. Base on these, would you please maintain two types dockerfile, one for release, the other for debug. I suggest to use centos as the base image for debug version, it is widely used by lots of company in production env.

**Comment by LuQQiu** [comment]
Centos docker image is even bigger than the ubuntu one. Should we offer the official testing docker image? If we should, a bigger Alluxio docker image based on either ubuntu or centos is acceptable and can be under alluxio-test/alluxio @apc999 what are your thoughts? @Xenorith what are your thoughts on centos vs ubuntu?

**Comment by Xenorith** [comment]
i would propose naming the larger image as `alluxio-dev` instead of `alluxio-test`
i do not have a preference between ubuntu vs centos; this might be a better question for @madanadit given he's more familiar with production environments. i personally use centos more often and am less familiar with ubuntu, but that shouldn't affect this decision

**Comment by apc999** [comment]
maybe still in [alluixo/alluxio](https://hub.docker.com/repository/docker/alluxio/alluxio) repo, but with different tag name . e.g, suffixed by `-dev` to indicate larger image size and more tools inside (based on different image). Let's discuss it tonight.

**Comment by LuQQiu** [comment]
Redirect this PR to target feature branch `ssz-docker`.
This PR will target to simplify the docker image based on `alpine`
Next PR will target to add a profile to support building a docker image based on `centos`. @maobaolong can you share what changes may be needed to change the alluxio docker image to be on `centos`? i guess many install package commands are different

File: integration/docker/Dockerfile (hunk 1 above)
Comment (ssz1997) [reply integration/docker/Dockerfile:58]: We could, but the installation and uninstallation have to be on the same layer, which means we will have a giant RUN command, i.e. install all packages, install libfuse, uninstall all packages, all under the same RUN. Do we want to do that?

**Comment by ssz1997** [comment]
@LuQQiu Uninstall some packages after their usage to further decrease the image size. The image can be gotten by `docker pull ssz1997/alluxio-alpine:latest`. The tarball used is the latest OS master tarball. PTAL, thanks!

**Comment by LuQQiu** [comment]
@apc999 PTAL, thanks

**Comment by ssz1997** [comment]
After some consideration and input from @Xenorith (Thanks Rico!), I feel like it doesn't worth to create a common Dockerfile-common for Dockerfile and Dockerfile-dev for the following reasons:
1. alpine image needs as few packages as possible, but centOS needs many packages. Also they use different package managements (alpine uses apk, and centOS uses yum), so the commands are quite different. Therefore, any installation involving packages, including installing libfuse, java, all the various tools, can't be put into the common dockerfile.
2. The commands for creating user are also different because alpine supports less commands. 
3. Even if we put the step that grants some directory permission to user Alluxio in Dockerfile-common, we have to copy them out  to Dockerfile and Dockerfile-dev. Anything in common that's new, we have to copy them out into the two Dockerfiles, which may still involve modify our Dockerfiles.
4. Other than these 3 steps, the Dockerfile doesn't do much. The ARG and ENV variables are not sharable across different Dockerfiles. So we don't really have much to put into Dockerfile-common.
@LuQQiu @apc999 Please let me know if you have any question.

**Comment by LuQQiu** [comment]
@ssz1997 @Xenorith Based on our discussion, is it possible to put the common code in a script and then use that script in two dockerfiles?

**Comment by ssz1997** [comment]
@LuQQiu 90% of the commands are actually different. The only common ones are grant permissions to Alluxio worker.

**Comment by LuQQiu** [comment]
The libfuse related commands and grant permissions commands are common, is it possible to put them into a common script and trigger it in Dockerfile, or how hard it is

**Comment by ssz1997** [comment]
@LuQQiu It's not hard. Now I think it does make sense to put these commands into a script. I'll work on it.

**Comment by ssz1997** [comment]
@LuQQiu I suggest we merge this PR and the centOS one first, and then I'll create the PR for the common script.

**Review by LuQQiu — APPROVED** [review]
LGTM, thanks for improving the docker image size!

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please

**Pull request merged**